### PR TITLE
Add ByteArray binding

### PR DESCRIPTION
**Pull request by elmarco**
Not exactly the most useful thing, but an interesting exercise during
Rust+GNOME hackfest :)

(The gir definitions were a bit too weird to auto-generate a satisfying binding)

Signed-off-by: Marc-André Lureau <marcandre.lureau@redhat.com>

### CODE REVIEW

File: src/byte_array.rs [hunk 1]
```diff
@@ -0,0 +1,264 @@
...
+}
+
+impl Default for ByteArray {
+    fn default() -> Self {
+        Self::new()
+    }
+}
+
+unsafe impl Send for ByteArray {}
+unsafe impl Sync for ByteArray {}
```
Comment (GuillaumeGomez) [inline src/byte_array.rs]: `ByteArray` is a thread-safe type?

Comment (elmarco) [reply src/byte_array.rs]: as much as GString, which also impl Send/Sync.

but only refcounting seems safe

Comment (sdroege) [reply src/byte_array.rs]: It would only be Send/Sync if it used normal Rust mutability rules and no reference counting

File: src/byte_array.rs [hunk 2]
```diff
@@ -0,0 +1,264 @@
...
+        unsafe { from_glib_full(glib_sys::g_byte_array_new()) }
+    }
+
+    pub fn with_capacity(size: usize) -> ByteArray {
+        unsafe { from_glib_full(glib_sys::g_byte_array_sized_new(size as u32)) }
+    }
+
+    pub fn into_gbytes(self) -> Bytes {
+        unsafe {
+            let ret = from_glib_full(glib_sys::g_byte_array_free_to_bytes(mut_override(
```
Comment (sdroege) [inline src/byte_array.rs]: Mixing `free()` and refcounting seems... suboptimal? Is `free` just calling `unref()`?

File: src/byte_array.rs [hunk 3]
```diff
@@ -0,0 +1,264 @@
...
+            glib_sys::g_byte_array_append(
+                self.to_glib_none().0,
+                bytes.as_ptr() as *const _,
+                bytes.len() as u32,
+            );
+        }
+        self
+    }
+
+    pub fn prepend<T: AsRef<[u8]>>(&self, data: T) -> &Self {
```
Comment (sdroege) [inline src/byte_array.rs]: You could here (and above) take the data by reference instead. There's no need to take ownership of it here as the byte array will copy the data

File: src/byte_array.rs (hunk 1 above)
Comment (elmarco) [reply src/byte_array.rs]: I should drop it then?

File: src/byte_array.rs [hunk 4]
```diff
@@ -0,0 +1,264 @@
...
+
+impl Deref for ByteArray {
+    type Target = [u8];
+
+    fn deref(&self) -> &[u8] {
+        unsafe {
+            let ptr = (*self.to_glib_none().0).data;
+            let len = (*self.to_glib_none().0).len as usize;
+            debug_assert!(!ptr.is_null() || len == 0);
+            slice::from_raw_parts(ptr as *const u8, len)
```
Comment (sdroege) [inline src/byte_array.rs]: This is UB, the pointer must never be null. If the length is 0, pass in mem::align_of::<u8>() or so :)

File: src/byte_array.rs (hunk 1 above)
Comment (sdroege) [reply src/byte_array.rs]: Yes, it's wrong currently

File: src/byte_array.rs (hunk 2 above)
Comment (elmarco) [reply src/byte_array.rs]: g_byte_array_free*() is doing somewhat unconventional things. If there are > 1 refs, it frees the internal buffer, not the GArray "wrapper" (see PRESERVE_WRAPPER, https://gitlab.gnome.org/GNOME/glib/blob/master/glib/garray.c#L357)

File: src/byte_array.rs (hunk 1 above)
Comment (elmarco) [reply src/byte_array.rs]: ok, removed

File: src/byte_array.rs (hunk 2 above)
Comment (sdroege) [reply src/byte_array.rs]: And then what happens if you use the wrapper? It would start with a new internal buffer?

Comment (elmarco) [reply src/byte_array.rs]: Yes, it does g_realloc() internally which alloc when given NULL.

File: src/byte_array.rs (hunk 4 above)
Comment (elmarco) [reply src/byte_array.rs]: no idea what you mean here :) from_raw_parts(NULL, 0) wouldn't work as expected?

File: src/byte_array.rs (hunk 2 above)
Comment (sdroege) [reply src/byte_array.rs]: Seems fine then I guess :)

File: src/byte_array.rs (hunk 4 above)
Comment (sdroege) [reply src/byte_array.rs]: See https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html
```
data must be non-null and aligned, even for zero-length slices.
```

Comment (elmarco) [reply src/byte_array.rs]: Something like that?
```
+            if ptr.is_null() {
+                ptr = NonNull::dangling().as_ptr();
+            }
```

Comment (sdroege) [reply src/byte_array.rs]: Yeah

**Pull request merged**